### PR TITLE
Modernize plugin, reduce set of allowed elements

### PR DESCRIPTION
**Pull request by daniel-beck**
This is a proposal to modernize the plugin:

* The Jenkins 2.121 dependency minimizes implied dependencies and is the oldest release line supported by Jenkins project infrastructure
* Update the library dependency to a recent one
* Remove the unused "ebay policy" code
* Replace the "Myspace policy" with a fairly small one based on library presets, allowing basic formatting.

Notably, this drops support for form elements (`label`, `form`, `button`, `input`, `select`, `option`, `textarea`), a few more obscure formatting options (`pre`, `cite`, `samp`, `center`), the `style` tag, and a number of attributes. I do not think the previous breadth of elements, especially those used to build forms, are a reasonable default to include in something like Jenkins.

It's a bit unfortunate that the artifact ID mentions the specific choice of rule set, but that cannot be helped.

### CODE REVIEW

File: src/test/java/hudson/markup/MyspacePolicyTest.java [hunk 1]
```diff
@@ -15,14 +15,14 @@
 public class MyspacePolicyTest extends Assert {
```
Comment (Wadeck) [inline src/test/java/hudson/markup/MyspacePolicyTest.java]: Should be renamed to `BasicPolicyTest`

In addition, the readme should be updated to remove the reference to Myspace.

File: src/test/java/hudson/markup/MyspacePolicyTest.java [hunk 2]
```diff
@@ -64,7 +64,7 @@ private void assertReject(String problematic, String input) {
     }
     
     private void assertSanitize(String expected, String input) {
-        assertEquals(expected,sanitize(input));
+        assertEquals(expected.replace('\'','\"'),sanitize(input));
     }
 
     private String sanitize(String input) {
```
Comment (Wadeck) [inline src/test/java/hudson/markup/MyspacePolicyTest.java]: Same refactoring as `RawHtmlMarkupFormatter` could apply here

File: pom.xml [hunk 3]
```diff
@@ -43,9 +43,9 @@
     </pluginRepositories>
     <dependencies>
         <dependency>
-            <groupId>org.kohsuke</groupId>
-            <artifactId>owasp-html-sanitizer</artifactId>
-            <version>r88</version>
+            <groupId>com.googlecode.owasp-java-html-sanitizer</groupId>
+            <artifactId>owasp-java-html-sanitizer</artifactId>
+            <version>20190610.1</version>
```
Comment (Wadeck) [inline pom.xml]: 😍

File: src/test/java/hudson/markup/MyspacePolicyTest.java [hunk 4]
```diff
@@ -15,14 +15,14 @@
...
+        assertSanitize("<a href='relative/link' rel='nofollow'>relative</a>", "<a href='relative/link'>relative</a>");
+        assertSanitize("<a href='mailto:kk&#64;kohsuke.org' rel='nofollow'>myself</a>", "<a href='mailto:kk&#64;kohsuke.org'>myself</a>");
         assertReject("javascript","<a href='javascript:alert(5)'>test</a>");
 
-        assertIntact("<img src='http://www.cloudbees.com'>");
-        assertIntact("<img src='relative/test.png'>");
-        assertIntact("<img src='relative/test.png'>");
+        assertIntact("<img src='http://www.cloudbees.com' />");
+        assertIntact("<img src='relative/test.png' />");
+        assertIntact("<img src='relative/test.png' />");
```
Comment (Wadeck) [inline src/test/java/hudson/markup/MyspacePolicyTest.java]:
```suggestion
        assertIntact("<img src='relative/test.png' />");
        assertReject("onerror","<img src='x' onerror='alert(5)'>");
```

Tested locally, it passes :)

File: src/test/java/hudson/markup/MyspacePolicyTest.java [hunk 5]
```diff
@@ -40,10 +40,10 @@ public void testPolicy() {
         assertReject("script","<script src='http://foo/evil.js'></script>");
         assertReject("script","<script src='relative.js'></script>");
 
-        assertIntact("<style>H1 { display:none; }</style>");
+        assertReject("style", "<style>H1 { display:none; }</style>");
```
Comment (Wadeck) [inline src/test/java/hudson/markup/MyspacePolicyTest.java]:
```suggestion
        assertReject("style", "<style>H1 { display:none; }</style>");
        assertReject("style", "<div style='background-color: expression(alert(123));'>inline CSS</div>");
```

Tested locally, it passes :)

**Comment by jglick** [comment]
Still planning to work on this?

**Comment by daniel-beck** [comment]
> Still planning to work on this?

Yes; probably needs to be a 2.0 version of the plugin though with a compat warning (renders content very differently), hence I haven't continued here yet.

File: pom.xml [hunk 6]
```diff
@@ -44,9 +45,9 @@
     </pluginRepositories>
     <dependencies>
         <dependency>
-            <groupId>org.kohsuke</groupId>
-            <artifactId>owasp-html-sanitizer</artifactId>
-            <version>r88</version>
+            <groupId>com.googlecode.owasp-java-html-sanitizer</groupId>
```
Comment (jglick) [inline pom.xml]: :tada:

File: src/main/resources/hudson/markup/Messages.properties [hunk 7]
```diff
@@ -0,0 +1 @@
+RawHtmlMarkupFormatter.DisplayName=Safe HTML
```
Comment (jglick) [inline src/main/resources/hudson/markup/Messages.properties]: trailing newline please

File: src/test/java/hudson/markup/BasicPolicyTest.java [hunk 8]
```diff
@@ -64,28 +62,22 @@ private void assertReject(String problematic, String input) {
     }
     
     private void assertSanitize(String expected, String input) {
-        assertEquals(expected,sanitize(input));
+        assertEquals(expected.replace('\'','\"'),sanitize(input));
     }
 
     private String sanitize(String input) {
         StringBuilder buf = new StringBuilder();
         HtmlStreamRenderer renderer = HtmlStreamRenderer.create(
```
Comment (jglick) [inline src/test/java/hudson/markup/BasicPolicyTest.java]: Could this not be shared with `src/main/`?

**Comment by jglick** [comment]
Looks like I am listed as an owner, so let me know when you want a release. Or file a request to add yourself to RPU if you prefer.

**Comment by daniel-beck** [comment]
I think this should be done now. If you agree @jglick I'd appreciate a release. Thanks!

**Comment by daniel-beck** [comment]
(Probably best to squash-merge, but up to you.)

File: src/test/java/hudson/markup/BasicPolicyTest.java [hunk 9]
```diff
@@ -66,18 +68,10 @@ private void assertSanitize(String expected, String input) {
...
-                x -> {
-                    throw new AssertionError(x);
-                }
-        );
-        HtmlSanitizer.sanitize(input, BasicPolicy.POLICY_DEFINITION.apply(renderer));
-        return buf.toString();
+        try {
+            return new RawHtmlMarkupFormatter(false).translate(input);
+        } catch (IOException ex) {
+            throw new AssertionError(ex);
```
Comment (jglick) [inline src/test/java/hudson/markup/BasicPolicyTest.java]: (or just let the helper method throw it up)

Comment (daniel-beck) [reply src/test/java/hudson/markup/BasicPolicyTest.java]: I had that first but we have so many helper methods in this class I went for changing not everything all over the place.

**Comment by jglick** [comment]
2.0 released.

**Comment by snago** [comment]
I see this in Jenkins Plugin Manager:
> Warning: the new version of this plugin claims to use a different settings format than the installed version. Jobs using this plugin may need to be reconfigured, and/or you may not be able to cleanly revert to the prior version without manually restoring old settings. Consult the plugin release notes for details.

The release notes have very little details.

What impact could this upgrade have? What potential problems should I be on the lookout for afterwards?

**Comment by daniel-beck** [comment]
@snago Sorry about that, bad timing. See https://github.com/jenkinsci/antisamy-markup-formatter-plugin/blob/add384097c543524129eeec4066691a5a7e91e4e/README.adoc for a better corresponding readme update.

**Comment by jimbasilio** [comment]
So by removing <pre> tags you've removed the ability to put ascii art in the jenkins System Message.

Any other alternative to this?

I know it's a small detail, but I'm betting MOST instances of jenkin usage had SOME type of ascii art in their headers to help differentiate different servers in an organization and their intent.

Thanks for all you do.

**Comment by jimbasilio** [comment]
Looks like I can still do inline styles:

`<div style="font-family: monospace;white-space: pre;">`
`.... pre styled things here ....`
`</div>`

**Comment by daniel-beck** [comment]
> betting MOST instances of jenkin usage had SOME type of ascii art in their headers

I only know of Simple Theme Plugin + custom CSS color theme per instance, never used ASCII art.

**Comment by tschechniker** [comment]
Hi There,

why did you drop the support for form / input? We are using this a lot to handle manual triggering of other projects for example to stage an artifact. The triggered projects are parameterized pipelines and the form will fill these parameters and start the build. So updating the plugin to 2.0 is currently a show stopper for us. Maybe kind of a configuration would be cool?

**Comment by daniel-beck** [comment]
I don't think it's a good idea to allow users with fairly low level permissions in Jenkins to create real looking UI that could trick admins into performing unintended actions.

If you want custom UI, write a plugin.

**Comment by jglick** [comment]
Agreed, this kind of thing can be done in a custom plugin, or in Pipeline with `input` + `build`, or in some external orchestration tooling, etc. Using a markup formatter to create a `form` is dangerous. If you have no security concerns, you can use the “anything goes” formatter.

**Comment by apottere** [comment]
@daniel-beck @jglick sorry to necro this issue, but I wasn't sure where else to reach out.  Let me know if there's a better place to post this.

I found this while trying to track down [JENKINS-62215](https://issues.jenkins-ci.org/browse/JENKINS-62215).  The `Active Choices` plugin relied on this plugin's sanitation to sanitize markup generated by non-privileged users/scripts, but one of the use-cases of that plugin was to be able to generate custom `input` fields in job parameters, [from the docs](https://plugins.jenkins.io/uno-choice/):
> By design, the values of Reactive Reference parameters are NOT passed to the build environment with one important exception. When the choice type is set to Formatted HTML or Formatted Hidden HTML and the HTML is an 'input' element the value can be passed to the build. See the 'Advanced Usage' section for additional instructions.

<hr />

> If you have no security concerns, you can use the “anything goes” formatter.

While I understand why the list of elements was restricted and why `form` elements _could_ be dangerous, it seems like there should be some middle ground between "completely safe" and "your users can XSS you".  Maybe add another formatter that's for safe html plus form elements?

> If you want custom UI, write a plugin.

While I could probably recreate our Active Choices parameters in a custom plugin, it seems much easier to add another option for allowing form elements in the markup formatter, since that behavior existed until this update.

**Comment by daniel-beck** [comment]
This won't be added again, as the use case for this plugin is sanitizing inline descriptions in Jenkins, and we no longer support form elements there.

Active Choices can choose a different provider for HTML sanitization, explicitly allowing the elements that plugin needs to work.

**Pull request merged**